src/workloads/docs/StateMachineCrudActor.yml

```diff
@@ -0,0 +1,159 @@
+SchemaVersion: 2018-07-01
+Owner: "@mongodb/product-perf"
+
+description: |
+  TODO: This Actor does not exist yet. It will be implemented in PERF-2516.
```

Comment:
This seems like a well-reasoned idea and a good starting point for the syntax.

I don't think this needs to extract/refactor Crud Actor code to be used in multiple Actors. Rather, I think something that encapsulates the "tell me which set of Operations to run now" would be a nice refactor seam.

```cpp
template<typename Op>
struct OpChooser {
    // Get list of operations to perform in this iteration of the Phase.
    std::vector<std::unique_ptr<Op>> ops(PhaseNumber currentPhase) const;
};

template<typename Op, typename...Args>
OpChooser<Op> make(
    const PhaseContext& ctx,
    std::initializer_list<std::pair<std::string, std::unique_ptr<Op>>> ops;
    ) {}

struct Operation {
    virtual void run() = 0;
    virtual ~Operation() = default;
};

struct OperationA : public Operation {
    void run() override {}
};
struct OperationB : public Operation {
    void run() override {}
};

struct PhaseConfig {
    OpChooser<Operation> chooser;
    PhaseConfig(const PhaseContext& ctx)
    : chooser{make<Operation>(
        ctx, {
            {"OperationA", std::make_unique<OperationA>()},
            {"OperationB", std::make_unique<OperationB>()}
        }
    )} {}
};

struct MyActor {
    void run() {
        for (auto&& config : _loop) {
            for (const auto&& op : config->chooser->ops()) {
                op->run()
            }
    }
    PhaseLoop<PhaseConfig> _loop;
};
```

Then the `OpsChooser` class knows how to translate the yaml structures of `Choice`/`Weight` and is constructed with the binding from operation names to impls. There is `Operations` xor `States` in the `Phase` block, so when it sees just `Operations`, the chooser is a static choice similar to what CrudActor currently does.

I'm dubious if this wants to be available to other Actors or if it wants to be isolated to CrudActor.{cpp,hpp} until its better proven out. I think it can easily start in just CrudActor and later brought out (or perhaps incorporated into PhaseLoop) if/when other Actors want similar functionality.

Also as mentioned in slack, there is reasonable fear/objection to changing CrudActor given its size. I suggest we move Crud Actor to its own cmake module as a peer with `cast_core` which will let us cleanly carve the cpp into multiple translation-units.

Reply:
@rtimmons would the only change here then be to make this "Crud" rather than "CrudState", or is there other syntax changes for it? I would be game for trying that. The code samples above scare me a bit, but I can deal and ask for help when I break the template stuff.

Reply:
@dalyd 

> would the only change here then be to make this "Crud" rather than "CrudState"

Yes that's the biggest thing. I have some 💅 things to think on for syntax keywords, but the structure seems good. I'm glad the transition table is enumerated per-state rather than global or path/context-dependent.

But actually two questions upon further thought:

1. When is the state transition logic evaluated? After every iteration of the Operations loop, or at some interval? If after every iteration, the number of transitions taken could be rather large, which is maybe what you want. If at some time or iteration interval, do we want syntax to make that explicit?
2. I'm confused about the `Delay` functionality in `Transitions`. According to your comment, "The delay is the time before firing the operations," but what's the distinction between this and the  "Sleep" state?

The "sleep"/"delay" things may be slightly trickier to implement without impacting performance or digging kinda deep into `Orchestrator` or `PhaseLoop` internals. We already have [SleepBefore](https://github.com/mongodb/genny/blob/5d20179b3dafd7ae2d63e35f26e5d7782490e16e/src/workloads/docs/HelloWorld.yml#L11-L14) which we may want to unify with this concept.

Reply:
Reverse order on responses (kindof): 

1. The sleep state was meant to be part of the model. I had a cell phone or laptop in mind that could go into a sleep state. It's completely separate from the FSM machinery. If it's confusing I can completely put in a different state there, or I could expand the doc at the top to explain it better. 
2. Delay: My model is in each state, we pick a transition. After picking the transition, we wait for whatever the delay is before transitioning to the next state. Repeat. There can be a lot of transitions (or not), and that's my goal. 
3. The order of delays and choices, and if delays are associated with the state or the transition are big ones. I went with what made sense to me, but there are other options.  
4. The delay can be seen as related to the rate limiting, but it feels different to me. It's definitely within a phase though. 

Also, please throw on the comments about the syntax keywords.

Reply:
Related to the delay question, this workload looks like it's performing a single defined set of operations, then sleeping before transitioning and performing the next defined set of operations. Would there be cases instead where instead we'd have a set of operations repeated, with either a `Duration` or `Repeat` keyword, before transitioning? Or maybe a number of repeats or durations with sleeps before/after?

Asking because that would seem to be replicating more `PhaseLoop` logic.

Reply:
> After picking the transition, we wait for whatever the delay is before transitioning

It sounds like `Duration` is the existing concept we could leverage.

And since we're getting `Duration` in States, I wonder if the rest of the Phase-level syntax keywords want to work here as well. I.e., `Duration`, `Repeat`, `SleepBefore`, rate-limiting, etc. That would certainly unify the syntax between Phases and States.

But it requires some nontrivial (though not huge) surgery to make it happen. I *think* this is more or less equivalent to baking the FSM machinery into PhaseLoop.

It gets hard to reason about rate-limiting and `Blocking` at the `Phase` level and at the `State` level, so we'd probably want to have some guidance and clear thinking on how to design with those.

Making the `Duration` value-generator-aware would be cool. We can clone `RandomDouble` or similar to make it a `ValueGenerator<TimeSpec>` to get your idea.

Here's my slight tweak on what you propose. Main differences (1) put Phase-like semantics into each `State`, and (2) use `WithProbability` as a percentage in favor of `Weights`. Last part is subjective, and first part has some pros/cons that are worth discussing.

```yaml
Actors:
- Name: Store
  Type: Crud
  Phases:
  - Duration: 10 minutes
    States:
    - Name: Find
      Duration:
        ^RandomDuration: {min: 1, max: 10, unit: seconds}
      GlobalRate: 1000 per second
      Operations: [{findOne}]
      Transitions:
      - To: Update
        WithProbability: 1%
      - To: {^Choice: [Update, Find]}
        WithProbability: 99%
    - Name: Update
      Repeat: 1000
      Operations: [{updateOne}]
      Transitions:
      - To: Find
        WithProbability: 10%
```

We could start with just `Duration` and reject other keywords for now. This would let us delay surgery to `PhaseLoop` at the expense of duplicating some [`TimeSpec` and time-check logic](https://github.com/mongodb/genny/blob/e8cc83e21b32a2713201207f8d8fe7624e0dd005/src/gennylib/include/gennylib/PhaseLoop.hpp#L170-L175).

Reply:
> Related to the delay question, this workload looks like it's performing a single defined set of operations, then sleeping before transitioning and performing the next defined set of operations. Would there be cases instead where instead we'd have a set of operations repeated, with either a `Duration` or `Repeat` keyword, before transitioning? Or maybe a number of repeats or durations with sleeps before/after?
> 
> Asking because that would seem to be replicating more `PhaseLoop` logic.

That is a great question. Not for my current case, but I could imagine cases in which it would be true. I'm thinking a lot about whether I want this or not.

Reply:
@rtimmons with your example, global rate doesn't make sense. I imagine it's a state per actor instance, as opposed to all the instantiations transitioning at the same time.